source/slang/slang-ir-metal-legalize.cpp

```diff
@@ -10,709 +10,1371 @@
+                    NestedFragment2 p3;
+                };
 
-        IRBuilder builder(func);
-        builder.setInsertBefore(func);
-        IRStructType* structType = builder.createStructType();
-        auto stageText = getStageText(entryPoint.entryPointDecor->getProfile().getStage());
-        builder.addNameHintDecoration(structType, (String(stageText) + toSlice("Input")).getUnownedSlice());
-        List<IRStructKey*> keys;
-        IRStructTypeLayout::Builder layoutBuilder(&builder);
-        for (auto param : paramsToPack)
-        {
-            auto paramVarLayout = findVarLayout(param);
-            auto key = builder.createStructKey();
-            param->transferDecorationsTo(key);
-            builder.createStructField(structType, key, param->getDataType());
-            if (auto varyingInOffsetAttr = paramVarLayout->findOffsetAttr(LayoutResourceKind::VaryingInput))
-            {
-                if (!key->findDecoration<IRSemanticDecoration>() && !paramVarLayout->findAttr<IRSemanticAttr>())
+                // Legalized with flattening -- current approach
+                struct Fragment
                 {
-                    // If the parameter doesn't have a semantic, we need to add one for semantic matching.
-                    builder.addSemanticDecoration(key, toSlice("_slang_attr"), (int)varyingInOffsetAttr->getOffset());
-                }
-            }
-            if (isGeometryStage)
+                    float4 p1 : SV_TARGET0;
+                    float2 p2 : SV_TARGET1;
+                    float2 p3 : SV_TARGET2;
+                };
+            */
+
+            auto func = entryPoint.entryPointFunc;
+            bool modified = false;
+            for (auto param : func->getParams())
             {
-                // For geometric stages, we need to translate VaryingInput offsets to MetalAttribute offsets.
-                IRVarLayout::Builder elementVarLayoutBuilder(&builder, paramVarLayout->getTypeLayout());
-                elementVarLayoutBuilder.cloneEverythingButOffsetsFrom(paramVarLayout);
-                for (auto offsetAttr : paramVarLayout->getOffsetAttrs())
+                auto layout = findVarLayout(param);
+                if (!layout)
+                    continue;
+                if (!layout->findOffsetAttr(LayoutResourceKind::VaryingInput))
+                    continue;
+                if (param->findDecorationImpl(kIROp_HLSLMeshPayloadDecoration))
+                    continue;
+                // If we find a IRParam with a IRStructType member, we need to flatten the entire IRParam
+                if (auto structType = as<IRStructType>(param->getDataType()))
                 {
-                    auto resourceKind = offsetAttr->getResourceKind();
-                    if (resourceKind == LayoutResourceKind::VaryingInput)
+                    IRBuilder builder(func);
+                    Dictionary<IRStructField*, IRStructField*> mapOldFieldToNewField;
```

Comment:
How can this dictionary not having conflicting keys if a struct type is used in duplicate fields? After flattening a single original field can map to different flattened fields, right?

Reply:
This is an oversight with the current design. 
The hierarchy of members needs to be tracked to solve this problem.